File: cisco-dashboard-frontend/src/types/policy.ts
/**
 * Policy types for Cisco SD-WAN Forensic Dashboard
 * These types mirror the backend models/policy.go structures
 */

/**
 * LocalPolicyEntry represents a single policy item running on a device.
 */
export interface LocalPolicyEntry {
  name: string;
  type: string;
  direction?: string;
  interface?: string;
  sequence?: string;
  cir?: string;
  burst?: string;
  exceedAction?: string;
  defaultAction?: string;
  action?: string;
}

/**
 * LocalPolicyResponse is the response from /api/device/{system-ip}/policy/local
 */
export interface LocalPolicyResponse {
  systemIp: string;
  hostName: string;
  siteId: string;
  accessLists: LocalPolicyEntry[];
  qosMaps: LocalPolicyEntry[];
  policers: LocalPolicyEntry[];
  zoneFirewall: LocalPolicyEntry[];
  totalCount: number;
}

/**
 * PolicySequenceInfo holds simplified sequence match/action info.
 */
export interface PolicySequenceInfo {
  sequenceName: string;
  sequenceType: string;
  baseAction: string;
  match?: Record<string, string>;
  actions?: Record<string, string>;
}

/**
 * CentralPolicyMatch represents a centralized policy that affects a device.
 */
export interface CentralPolicyMatch {
  policyName: string;
  policyId: string;
  policyType: string;
  isActive: boolean;
  sequences?: PolicySequenceInfo[];
}

/**
 * CentralPolicyResponse is the response from /api/device/{system-ip}/policy/centralized
 */
export interface CentralPolicyResponse {
  systemIp: string;
  hostName: string;
  siteId: string;
  dataPolicies: CentralPolicyMatch[];
  controlPolicies: CentralPolicyMatch[];
  appRoutePolicies: CentralPolicyMatch[];
  totalCount: number;
}

/**
 * Friendly labels for match/action field names
 */
export const FIELD_LABELS: Record<string, string> = {
  sourceDataPrefixList: "Source Prefix",
  destinationDataPrefixList: "Dest Prefix",
  sourceIp: "Source IP",
  destinationIp: "Dest IP",
  sourcePort: "Source Port",
  destinationPort: "Dest Port",
  protocol: "Protocol",
  dscp: "DSCP",
  app: "Application",
  appList: "App List",
  dnsAppList: "DNS App",
  dns: "DNS",
  packetLength: "Pkt Length",
  plp: "PLP",
  trafficTo: "Traffic To",
  localTlocColor: "Local TLOC Color",
  preferredColorGroup: "Preferred Color",
  set: "Set",
  nat: "NAT",
  redirect: "Redirect",
  log: "Log",
  count: "Counter",
  cflowd: "Cflowd",
  tcpOptimization: "TCP Opt",
  lossCorrection: "FEC",
  sig: "SIG Redirect",
};

/**
 * Format a field name to a human-readable label
 */
export function formatFieldLabel(field: string): string {
  if (FIELD_LABELS[field]) {
    return FIELD_LABELS[field];
  }
  // Convert camelCase to Title Case
  return field
    .replace(/([A-Z])/g, " $1")
    .replace(/^./, (s) => s.toUpperCase());
}

/**
 * Determine if an action is a drop/deny action
 */
export function isDropAction(action: string): boolean {
  const lower = (action || "").toLowerCase();
  return lower === "drop" || lower === "reject" || lower === "deny";
}

/**
 * Determine if an action is an accept/permit action
 */
export function isAcceptAction(action: string): boolean {
  const lower = (action || "").toLowerCase();
  return lower === "accept" || lower === "forward" || lower === "permit" || lower === "pass";
}

File: models/policy.go
package models

// LocalPolicyEntry represents a single policy item running on a device.
type LocalPolicyEntry struct {
	Name         string `json:"name"`
	Type         string `json:"type"`
	Direction    string `json:"direction,omitempty"`
	Interface    string `json:"interface,omitempty"`
	Sequence     string `json:"sequence,omitempty"`
	CIR          string `json:"cir,omitempty"`
	Burst        string `json:"burst,omitempty"`
	ExceedAction string `json:"exceedAction,omitempty"`
}

// LocalPolicyResponse is the consolidated response for local policy data.
type LocalPolicyResponse struct {
	SystemIP     string             `json:"systemIp"`
	HostName     string             `json:"hostName"`
	SiteID       string             `json:"siteId"`
	AccessLists  []LocalPolicyEntry `json:"accessLists"`
	QosMaps      []LocalPolicyEntry `json:"qosMaps"`
	Policers     []LocalPolicyEntry `json:"policers"`
	ZoneFirewall []LocalPolicyEntry `json:"zoneFirewall"`
	TotalCount   int                `json:"totalCount"`
}

// PolicySequenceInfo holds simplified sequence match/action info.
type PolicySequenceInfo struct {
	SequenceName string            `json:"sequenceName"`
	SequenceType string            `json:"sequenceType"`
	BaseAction   string            `json:"baseAction"`
	Match        map[string]string `json:"match,omitempty"`
	Actions      map[string]string `json:"actions,omitempty"`
}

// CentralPolicyMatch represents a centralized policy that affects a device.
type CentralPolicyMatch struct {
	PolicyName string               `json:"policyName"`
	PolicyID   string               `json:"policyId"`
	PolicyType string               `json:"policyType"`
	IsActive   bool                 `json:"isActive"`
	Sequences  []PolicySequenceInfo `json:"sequences,omitempty"`
}

// CentralPolicyResponse is the consolidated response for centralized policy data.
type CentralPolicyResponse struct {
	SystemIP        string               `json:"systemIp"`
	HostName        string               `json:"hostName"`
	SiteID          string               `json:"siteId"`
	DataPolicies    []CentralPolicyMatch `json:"dataPolicies"`
	ControlPolicies []CentralPolicyMatch `json:"controlPolicies"`
	AppPolicies     []CentralPolicyMatch `json:"appRoutePolicies"`
	TotalCount      int                  `json:"totalCount"`
}

// PolicyListInfo holds the resolved name and sample entries for a policy list.
type PolicyListInfo struct {
	Name    string   `json:"name"`
	Entries []string `json:"entries"` // first few entries for preview
}

// VSmartPolicy represents a vSmart policy from vManage.
type VSmartPolicy struct {
	PolicyID          string `json:"policyId"`
	PolicyName        string `json:"policyName"`
	PolicyType        string `json:"policyType"`
	IsPolicyActivated bool   `json:"isPolicyActivated"`
	PolicyDefinition  string `json:"policyDefinition"`
}
